workflow/workflow.go

```diff
@@ -262,6 +262,24 @@ func SideEffect(ctx Context, f func(ctx Context) interface{}) encoded.Value {
 	return internal.SideEffect(ctx, f)
 }
 
+// MutableSideEffect executes provided function once, then it look up history for value with the given id,
+// if there is no existing value, then it record the function result as value with the given id on history;
+// otherwise, it compare if the existing value from history has changed from the new function result by calling the
+// provided equals function,
+//   if they are equal, it returns the value without recording a new one in history;
+//   otherwise, it records the new value with the same id on history.
+//
+// Caution: do not use MutableSideEffect to modify closures, always retrieve result from MutableSideEffect's encoded
+// return value.
+//
+// The difference between MutableSideEffect() and SideEffect() is that every new SideEffect() call in non-replay will
+// result in a new marker been recorded on history. However, MutableSideEffect() only record a new marker if the value
+// changed. During replay, MutableSideEffect() will not execute the function again, but it will return the exact same
+// value as it was returning during the non-replay run.
+func MutableSideEffect(ctx Context, id string, f func(ctx Context) interface{}, equals func(a, b encoded.Value) bool) encoded.Value {
```

Comment:
I'm not sure if it is worth the optimization, but you could save the unnecessary  serialization of function result if equals function would have the following signature:

func(a interface{}, b encoded.Value) bool

Reply:
Actually I think you can even have func(a, b interface{}) as a has a type and it can be used to decode b.